File: f-sharp/ProductExport/Tests/XmlExporterTests.fs
[<VerifyXunit.UsesVerify>]
module ``Product Export``

open System
open System.Text
open System.Text.RegularExpressions
open System.Xml
open FSharp.Dependency.ProductExport
open FSharp.ProductExport.Tests
open VerifyTests
open VerifyXunit
open Xunit

let defaultSettings = VerifySettings()
defaultSettings.UseDirectory "./Data"
defaultSettings.UseExtension "xml"

let verifyXml xml settings =
    task {
        let xmlDoc = XmlDocument()
        xmlDoc.LoadXml xml
        return! Verifier.Verify (xmlDoc, settings)
    }

let toCSList (input : 'a list) =
    ResizeArray<'a> input

[<Fact>]
let ``tax details`` () =
    let xml = XmlExporter.ExportTaxDetails ([ SampleObjects.recentOrder ; SampleObjects.oldOrder ] |> toCSList)
    verifyXml xml defaultSettings

[<Fact>]
let ``store`` () =
    let xml = XmlExporter.ExportStore SampleObjects.flagshipStore
    verifyXml xml defaultSettings

[<Fact>]
let ``history`` () =
    let xml = XmlExporter.ExportHistory ([ SampleObjects.recentOrder ; SampleObjects.oldOrder ] |> toCSList)
    let settings = defaultSettings
    let scrubber (input: StringBuilder) =
        let regex = "createdAt=\"[^\"]+\""
        let replacement = "createdAt=\"2018-09-20T00:00Z\""
        let scrubbed = Regex.Replace (input.ToString(), regex, replacement)
        input.Clear().Append(scrubbed) |> ignore
    settings.AddScrubber scrubber
    verifyXml xml settings

[<Fact>]
let ``full`` () =
    let xml = XmlExporter.ExportFull ([ SampleObjects.recentOrder ; SampleObjects.oldOrder ] |> toCSList)
    verifyXml xml defaultSettings

File: c-sharp/ProductExport/Tests/XmlExporterTests.cs
using System.Text;
using System.Text.RegularExpressions;
using System.Xml;
using CSharp.ProductExport.Implementation;

namespace CSharp.ProductExport.Tests;

[UsesVerify]
public class XmlExporterTests
{
    [Fact]
    public Task ExportTaxDetails()
    {
        var orders = new List<Order> {SampleModelObjects.RecentOrder, SampleModelObjects.OldOrder};
        var xml = XmlExporter.ExportTaxDetails(orders);

        return VerifyXml(xml, DefaultSettings);
    }

    [Fact]
    public Task ExportStore()
    {
        var store = SampleModelObjects.FlagshipStore;
        var xml = XmlExporter.ExportStore(store);
        return VerifyXml(xml, DefaultSettings);
    }

    [Fact]
    public Task ExportHistory()
    {
        var orders = new List<Order> {SampleModelObjects.RecentOrder, SampleModelObjects.OldOrder};
        var xml = XmlExporter.ExportHistory(orders);
        var settings = DefaultSettings;
        settings.AddScrubber(
            input =>
            {
                var regex = "createdAt=\"[^\"]+\"";
                var replacement = "createdAt=\"2018-09-20T00:00Z\"";
                var scrubbed = Regex.Replace(input.ToString(), regex, replacement);
                input.Clear().Append(scrubbed);
            });
        return VerifyXml(xml, settings);
    }

    private static VerifySettings DefaultSettings
    {
        get
        {
            var verifySettings = new VerifySettings();
            verifySettings.UseDirectory("./Data");
            verifySettings.UseExtension("xml");
            return verifySettings;
        }
    }

    private static Task VerifyXml(string xml, VerifySettings settings)
    {
        try
        {
            var xmlDoc = new XmlDocument();
            xmlDoc.LoadXml(xml);
            return Verify(ToIndentedString(xmlDoc), settings);
        }
        catch (Exception e)
        {
            return Verify(xml);
        }
    }

    static string ToIndentedString(XmlDocument doc)
    {
        var stringWriter = new StringWriter(new StringBuilder());
        var xmlTextWriter = new XmlTextWriter(stringWriter) {Formatting = Formatting.Indented};
        doc.Save(xmlTextWriter);
        return stringWriter.ToString();
    }
}